tracee/tracee.go

```diff
@@ -0,0 +1,487 @@
+type context struct {
+	Ts      uint64   `json:"ts"`
+	Pid     uint32   `json:"pid"`
+	Tid     uint32   `json:"tid"`
+	Ppid    uint32   `json:"ppid"`
+	Uid     uint32   `json:"uid"`
+	MntId   uint32   `json:"mnt_id"`
+	PidId   uint32   `json:"ppid_id"`
+	Comm    taskComm `json:"comm"`
+	UtsName taskComm `json:"uts_name"`
+	Eventid uint32   `json:"eventid"`
+	Argnum  uint8    `json:"argnum"`
+	_       [3]byte  // padding for Argnum
+	Retval  int64    `json:"retval"`
+}
+
+// TraceeConfig is a struct containing user defined configuration of tracee
+type TraceeConfig struct {
+	Syscalls              map[string]bool
+	Sysevents             map[string]bool
+	ContainerMode         bool
+	DetectOriginalSyscall bool
+	OutputFormat          string
+}
+
+// Validate does static validation of the configuration
+func (tc TraceeConfig) Validate() error {
+	if tc.Syscalls == nil || tc.Sysevents == nil {
+		return fmt.Errorf("trace config validation failed: sysevents or syscalls is nil")
+	}
+	if tc.OutputFormat != "table" && tc.OutputFormat != "json" {
+		return fmt.Errorf("trace config validation failed: unrecognized output format: %s", tc.OutputFormat)
+	}
+	for sc, wanted := range tc.Syscalls {
+		_, valid := Syscalls[sc]
+		if wanted && !valid {
+			return fmt.Errorf("invalid syscall to trace: %s", sc)
+		}
+	}
+	for se, wanted := range tc.Sysevents {
+		_, valid := Sysevents[se]
+		if wanted && !valid {
+			return fmt.Errorf("invalid sysevent to trace: %s", se)
+		}
+	}
+	return nil
+}
+
+// NewConfig creates a new TraceeConfig instance based on the given configuration
+func NewConfig(eventsToTrace []string, containerMode bool, detectOriginalSyscall bool, outputFormat string) (*TraceeConfig, error) {
+	//separate eventsToTrace into syscalls and sysevents
+	syscalls := make(map[string]bool)
+	sysevents := make(map[string]bool)
+	if eventsToTrace == nil {
+		for sc := range Syscalls {
+			eventsToTrace = append(eventsToTrace, sc)
+		}
+		for se := range Sysevents {
+			eventsToTrace = append(eventsToTrace, se)
+		}
+	}
+	for _, e := range eventsToTrace {
+		if enabled, ok := Syscalls[e]; ok {
+			syscalls[e] = enabled
+		}
+		if enabled, ok := Sysevents[e]; ok {
+			sysevents[e] = enabled
+		}
+	}
+
+	tc := TraceeConfig{
+		Syscalls:              syscalls,
+		Sysevents:             sysevents,
+		ContainerMode:         containerMode,
+		DetectOriginalSyscall: detectOriginalSyscall,
+		OutputFormat:          outputFormat,
+	}
+
+	err := tc.Validate()
+	if err != nil {
+		return nil, err
+	}
+	return &tc, nil
+}
+
+// Tracee traces system calls and events using eBPF
+type Tracee struct {
+	config         TraceeConfig
+	bpfProgramPath string
+	bpfModule      *bpf.Module
+	bpfPerfMap     *bpf.PerfMap
+	eventsChannel  chan []byte
+	printer        eventPrinter
+}
+
+// New creates a new Tracee instance based on a given valid TraceeConfig
+func New(cfg TraceeConfig) (*Tracee, error) {
+	var err error
+
+	// validation
+	err = cfg.Validate()
```

Comment:
The state of mind here is that tracee is a library. The part of creating the configuration is a distinct operation from running tracee, and at this point we can already fail early if the input is user provided invalid. The `New` function is the entry point for the library, and it should be able to care for itself and validate input regardless of if a user validated before. They might create the config manually instead of using the `NewConfig` helper, or change the config before passing it to `New`.